presto-common/src/main/java/com/facebook/presto/common/type/Type.java

```diff
@@ -186,4 +189,36 @@ public interface Type
      * Compare the values in the specified block at the specified positions equal.
      */
     int compareTo(Block leftBlock, int leftPosition, Block rightBlock, int rightPosition);
+
+    /**
+     * Returns the maximum value that compares less than {@code value}.
+     * <p>
+     * The type of the value must match {@link #getJavaType}.
+     *
+     * @throws IllegalStateException if this type is not {@link #isOrderable() orderable}
+     */
+    default Optional<Object> getPreviousValue(Object value)
```

Comment:
Just for consistency, should we implement these for other orderable types such as Double, Real, etc?

Reply:
Also, because Types may be registered through the SPI, I think instead of returning optional empty, by default it should throw.  All engine recognized types that are orderable should have implementations, and SPI types that don't implement this should throw clear exceptions that indicate to plugin authors what to do to fix the problem.

Reply:
> Just for consistency, should we implement these for other orderable types such as Double, Real, etc?

For now we only used this method for Iceberg's non-identity/bucket/void transform, so the implementation of double/real are useless at the moment, as they are not the supported types for truncate/year/month/day/hour. But as your suggestion, we could support double/real based on java provided method `Math.ulp()` for consistency and potential future usage. And I will figure out whether we can also support decimal.

Reply:
> Also, because Types may be registered through the SPI, I think instead of returning optional empty, by default it should throw. All engine recognized types that are orderable should have implementations, and SPI types that don't implement this should throw clear exceptions that indicate to plugin authors what to do to fix the problem.

Good suggestion, make the default methods in `Type` to throw and let individual types to determine whether it could figure out the previous value and next value. In that way, all engine recognized types that are orderable implement the two methods and return corresponding result or an empty optional to declare that  it could not figure out the value exactly adjacent a specified value. And the SPI type would be reminded clearly. I will fix it.